flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SourceTransformationTranslator.java

```diff
@@ -36,26 +38,29 @@
  * @param <OUT> The type of the elements that this source produces.
  */
 @Internal
-public class SourceTransformationTranslator<OUT>
-		extends SimpleTransformationTranslator<OUT, SourceTransformation<OUT>> {
+public class SourceTransformationTranslator<OUT, SplitT extends SourceSplit, EnumChkT>
+		extends SimpleTransformationTranslator<OUT, SourceTransformation<OUT, SplitT, EnumChkT>> {
 
 	@Override
 	protected Collection<Integer> translateForBatchInternal(
-			final SourceTransformation<OUT> transformation,
+			final SourceTransformation<OUT, SplitT, EnumChkT> transformation,
 			final Context context) {
-		return translateInternal(transformation, context);
+
+		return translateInternal(transformation, context, false /* emit progressive watermarks */);
 	}
 
 	@Override
 	protected Collection<Integer> translateForStreamingInternal(
-			final SourceTransformation<OUT> transformation,
+			final SourceTransformation<OUT, SplitT, EnumChkT> transformation,
 			final Context context) {
-		return translateInternal(transformation, context);
+
+		return translateInternal(transformation, context, true /* emit progressive watermarks */);
 	}
 
 	private Collection<Integer> translateInternal(
-			final SourceTransformation<OUT> transformation,
-			final Context context) {
+			final SourceTransformation<OUT, SplitT, EnumChkT> transformation,
+			final Context context,
+			boolean emitProgressiveWatermarks) {
```

Comment:
Why not simply `emitWatermarks`?

Reply:
It is to show that we emit watermarks as we go. Even in BATCH mode we still emit the final watermark.